paz/models/detection/model_util.py

```diff
@@ -0,0 +1,268 @@
+    row2 = tf.stack([zeros, ones, zeros], axis=1)
+    row3 = tf.stack([-s, zeros, c], axis=1)
+    R = tf.concat([tf.expand_dims(row1, 1), tf.expand_dims(row2, 1), tf.expand_dims(row3, 1)], axis=1)  # (N,3,3)
+    # print row1, row2, row3, R, N
+    corners_3d = tf.matmul(R, corners)  # (N,3,8)
+    corners_3d += tf.tile(tf.expand_dims(centers, 2), [1, 1, 8])  # (N,3,8)
+    corners_3d = tf.transpose(corners_3d, perm=[0, 2, 1])  # (N,8,3)
+    return corners_3d
+
+
+def get_box3d_corners(center, heading_residuals, size_residuals):
+    """ TF layer.
+    Inputs:
+        center: (B,3)
+        heading_residuals: (B,NH)
+        size_residuals: (B,NS,3)
+    Outputs:
+        box3d_corners: (B,NH,NS,8,3) tensor
+    """
+    batch_size = center.get_shape()[0]
+    heading_bin_centers = tf.constant(np.arange(0, 2 * np.pi, 2 * np.pi / NUM_HEADING_BIN), dtype=tf.float32)  # (NH,)
+    headings = heading_residuals + tf.expand_dims(heading_bin_centers, 0)  # (B,NH)
+
+    mean_sizes = tf.expand_dims(tf.constant(g_mean_size_arr, dtype=tf.float32), 0) + size_residuals  # (B,NS,1)
+    sizes = mean_sizes + size_residuals  # (B,NS,3)
+    sizes = tf.tile(tf.expand_dims(sizes, 1), [1, NUM_HEADING_BIN, 1, 1])  # (B,NH,NS,3)
+    headings = tf.tile(tf.expand_dims(headings, -1), [1, 1, NUM_SIZE_CLUSTER])  # (B,NH,NS)
+    centers = tf.tile(tf.expand_dims(tf.expand_dims(center, 1), 1),
+                      [1, NUM_HEADING_BIN, NUM_SIZE_CLUSTER, 1])  # (B,NH,NS,3)
+
+    N = batch_size * NUM_HEADING_BIN * NUM_SIZE_CLUSTER
+    corners_3d = get_box3d_corners_helper(tf.reshape(centers, [N, 3]), tf.reshape(headings, [N]),
+                                          tf.reshape(sizes, [N, 3]))
+
+    return tf.reshape(corners_3d, [batch_size, NUM_HEADING_BIN, NUM_SIZE_CLUSTER, 8, 3])
+
+
+def huber_loss(error, delta):
+    abs_error = tf.abs(error)
+    quadratic = tf.minimum(abs_error, delta)
+    linear = (abs_error - quadratic)
+    losses = 0.5 * quadratic ** 2 + delta * linear
+    return tf.reduce_mean(losses)
+
+
+def parse_output_to_tensors(output, end_points):
+    """ Parse batch output to separate tensors (added to end_points)
+    Input:
+        output: TF tensor in shape (B,3+2*NUM_HEADING_BIN+4*NUM_SIZE_CLUSTER)
+        end_points: dict
+    Output:
+        end_points: dict
+    """
+    batch_size = output.get_shape()[0]
+    center = tf.slice(output, [0, 0], [-1, 3])
+    end_points['center_boxnet'] = center
+
+    heading_scores = tf.slice(output, [0, 3], [-1, NUM_HEADING_BIN])
+    heading_residuals_normalized = tf.slice(output, [0, 3 + NUM_HEADING_BIN],
+                                            [-1, NUM_HEADING_BIN])
+    end_points['heading_scores'] = heading_scores  # BxNUM_HEADING_BIN
+    end_points['heading_residuals_normalized'] = \
+        heading_residuals_normalized  # BxNUM_HEADING_BIN (-1 to 1)
+    end_points['heading_residuals'] = \
+        heading_residuals_normalized * (np.pi / NUM_HEADING_BIN)  # BxNUM_HEADING_BIN

```

Comment:
Maybe it would be easier to do all your calculations and at the end of the function (before returning the value) you create the dictionary. It just seems very difficult to read when you are entering the dictionary every single time.